rpc-server/src/modules/blocks/methods.rs

```diff
@@ -1,51 +1,253 @@
+        fetch_changes_in_block_by_type(&data, block, &params.state_changes_request).await;
+
+    #[cfg(feature = "shadow_data_consistency")]
+    {
+        let near_rpc_client = data.near_rpc_client.clone();
+        if let near_primitives::types::BlockReference::Finality(_) = params.block_reference {
+            params.block_reference = near_primitives::types::BlockReference::from(
+                near_primitives::types::BlockId::Height(block.block_height),
+            )
+        }
+        let error_meta = format!("CHANGES_IN_BLOCK_BY_TYPE: {:?}", params);
+        let read_rpc_response_json = match &changes_response {
+            Ok(res) => serde_json::to_value(res),
+            Err(err) => serde_json::to_value(err),
+        };
+        let comparison_result =
+            shadow_compare_results(read_rpc_response_json, near_rpc_client, params).await;
+
+        match comparison_result {
+            Ok(_) => {
+                tracing::info!(target: "shadow_data_consistency", "Shadow data check: CORRECT\n{}", error_meta);
+            }
+            Err(err) => {
+                tracing::warn!(target: "shadow_data_consistency", "Shadow data check: ERROR\n{}\n{:?}", error_meta, err);
+                crate::metrics::CHNGES_IN_BLOCK_BY_TYPE_PROXIES_TOTAL.inc()
+            }
+        }
+    }
+
+    Ok(changes_response.map_err(near_jsonrpc_primitives::errors::RpcError::from)?)
+}
+
 #[cfg_attr(feature = "tracing-instrumentation", tracing::instrument(skip(data)))]
 pub async fn fetch_block(
     data: &Data<ServerContext>,
     block_reference: near_primitives::types::BlockReference,
-) -> anyhow::Result<near_primitives::views::BlockView> {
+) -> Result<
+    near_jsonrpc_primitives::types::blocks::RpcBlockResponse,
+    near_jsonrpc_primitives::types::blocks::RpcBlockError,
+> {
     tracing::debug!("`fetch_block` call");
     let block_height = match block_reference {
         near_primitives::types::BlockReference::BlockId(block_id) => match block_id {
-            near_primitives::types::BlockId::Height(block_height) => block_height,
+            near_primitives::types::BlockId::Height(block_height) => Ok(block_height),
             near_primitives::types::BlockId::Hash(block_hash) => {
                 scylla_db_convert_block_hash_to_block_height(&data.scylla_db_manager, block_hash)
-                    .await?
+                    .await
             }
         },
         near_primitives::types::BlockReference::Finality(finality) => match finality {
-            near_primitives::types::Finality::Final => data
+            near_primitives::types::Finality::Final => Ok(data
                 .final_block_height
-                .load(std::sync::atomic::Ordering::SeqCst),
-            _ => anyhow::bail!("Finality other than final is not supported"),
+                .load(std::sync::atomic::Ordering::SeqCst)),
+            _ => Err(
+                near_jsonrpc_primitives::types::blocks::RpcBlockError::UnknownBlock {
```

Comment:
In our case, it's `InternalError` I think. This applies to the below cases as well.